contracts/reserve/GoodReserveCDai.sol

```diff
@@ -277,25 +355,105 @@ contract GoodReserveCDai is
 	 * @param _sellTo The tokens that will be received after the conversion
 	 * @param _gdAmount The amount of GD tokens that should be converted to `_sellTo` tokens
 	 * @param _minReturn The minimum allowed `sellTo` tokens return
+	 * @param _minTokenAmount The mininmum dai out amount from Exchange swap function
+	 * @param _targetAddress address of _sellTo token recipient if different than msg.sender
 	 * @return (tokenReturn) How much `sellTo` tokens were transferred
 	 */
 	function sell(
 		ERC20 _sellTo,
 		uint256 _gdAmount,
-		uint256 _minReturn
+		uint256 _minReturn,
+		uint256 _minTokenAmount,
+		address _targetAddress
 	) public returns (uint256) {
+		
+		address cDaiAddress = nameService.getAddress("CDAI");
+		address daiAddress = nameService.getAddress("DAI");
+		address receiver = _targetAddress == address(0x0) ? msg.sender : _targetAddress;
+		// redeclarations to prevent stack too deep errors
+		ERC20 tempSellTo = _sellTo;
+		uint256 tempMinReturn = _minReturn;
+		uint256 tempGdAmount = _gdAmount;
+		uint tempMinTokenAmount = _minTokenAmount;
+		uint256 result;
+		uint256 contributionAmount;
+		if(address(_sellTo) == cDaiAddress || address(_sellTo) == daiAddress){
+			(result,contributionAmount) = _sell(_gdAmount, _minReturn, address(tempSellTo) == cDaiAddress);
+			tempMinReturn = 0;
+			
+		} else {
+			uint256 returnAmount;
+			(returnAmount,contributionAmount) = _sell(_gdAmount, _minReturn, false);
+			address[] memory path = new address[](2);
+			path[0] = daiAddress;
+			path[1] = address(tempSellTo);
+			Uniswap uniswapContract = Uniswap(nameService.getAddress("UNISWAP_ROUTER"));
+			ERC20(daiAddress).approve(address(uniswapContract), returnAmount);
+			uint256[] memory swap = uniswapContract.swapExactTokensForTokens(
+					returnAmount,
+					tempMinTokenAmount,
+					path,
+					address(this),
+					block.timestamp
+				);
+			tempMinReturn = tempMinTokenAmount;
+			result = swap[1];
+			require(result > 0, "token selling failed");
+			
+			
+			
+		}
+		require(
+			result >= tempMinReturn,
+			"Token return must be above the minReturn"
+		);
+		require(
+			tempSellTo.transfer(receiver, result) == true,
+			"Transfer failed"
+		);
+		emit TokenSold(
```

Comment:
token sold is emitted in public method and tempSellTo is dynamic
while tokenpurchased is emitted in internal _buy and buyWith is always cdai and not dynamic